src/ctap/pin_protocol.rs

```diff
@@ -0,0 +1,460 @@
+        Ok(())
+    } else {
+        Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
+    }
+}
+
+pub struct SharedSecretV1 {
+    common_secret: [u8; 32],
+}
+
+impl SharedSecretV1 {
+    /// Creates a new shared secret from the handshake result.
+    fn new(handshake: [u8; 32]) -> SharedSecretV1 {
+        SharedSecretV1 {
+            common_secret: Sha256::hash(&handshake),
+        }
+    }
+
+    /// Creates a new shared secret for testing.
+    #[cfg(test)]
+    pub fn new_test(hash: [u8; 32]) -> SharedSecretV1 {
+        SharedSecretV1 {
+            common_secret: hash,
+        }
+    }
+}
+
+impl SharedSecret for SharedSecretV1 {
+    fn encrypt(
+        &mut self,
+        rng: &mut dyn Rng256,
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        encrypt(rng, &self.common_secret, plaintext, false)
+    }
+
+    fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Ctap2StatusCode> {
+        decrypt(&self.common_secret, ciphertext, false)
+    }
+
+    fn verify(&mut self, message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+        verify_v1(&self.common_secret, message, signature)
+    }
+}
+
+fn verify_v2(key: &[u8], message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+    if signature.len() != 32 {
+        return Err(Ctap2StatusCode::CTAP1_ERR_INVALID_PARAMETER);
+    }
+    if verify_hmac_256::<Sha256>(key, message, array_ref![signature, 0, 32]) {
+        Ok(())
+    } else {
+        Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID)
+    }
+}
+
+pub struct SharedSecretV2 {
+    handshake: [u8; 32],
+    aes_key: Option<[u8; 32]>,
+    hmac_key: Option<[u8; 32]>,
+}
+
+impl SharedSecretV2 {
+    /// Creates a new shared secret from the handshake result.
+    fn new(handshake: [u8; 32]) -> SharedSecretV2 {
+        SharedSecretV2 {
+            handshake,
+            aes_key: None,
+            hmac_key: None,
+        }
+    }
+
+    fn get_aes_key(&mut self) -> &[u8; 32] {
+        let handshake = &self.handshake;
+        self.aes_key
+            .get_or_insert_with(|| hkdf_empty_salt_256::<Sha256>(handshake, b"CTAP2 AES key"))
+    }
+
+    fn get_hmac_key(&mut self) -> &[u8; 32] {
+        let handshake = &self.handshake;
+        self.hmac_key
+            .get_or_insert_with(|| hkdf_empty_salt_256::<Sha256>(handshake, b"CTAP2 HMAC key"))
+    }
+}
+
+impl SharedSecret for SharedSecretV2 {
+    fn encrypt(
+        &mut self,
+        rng: &mut dyn Rng256,
+        plaintext: &[u8],
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        encrypt(rng, self.get_aes_key(), plaintext, true)
+    }
+
+    fn decrypt(&mut self, ciphertext: &[u8]) -> Result<Vec<u8>, Ctap2StatusCode> {
+        decrypt(self.get_aes_key(), ciphertext, true)
+    }
+
+    fn verify(&mut self, message: &[u8], signature: &[u8]) -> Result<(), Ctap2StatusCode> {
+        verify_v2(self.get_hmac_key(), message, signature)
```

Comment:
Why this calling indirection on both protocols instead of directly putting the code from the corresponding `verify_*` methods instead?
For encryption/decryption this makes sense from a factorisation point of view but that doesn't hold for `verify`

Reply:
Because `verify_pin_uv_auth_token` reuses them.

Reply:
Right, I missed that part.